AU2/plugins/util/ScoreManager.py

```diff
@@ -64,8 +70,38 @@ def _score(self, identifier: str) -> float:
         b = self._bonus(identifier)
         a = self._attempts(identifier)
         import math
-        score = eval(self.formula)
+        score = eval(self.formula) if self.formula else c
         return score
 
+    @functools.cache
+    def _rating(self, identifier: str) -> float:
+        game_end = self.game_end or get_now_dt()
+        deaths = self.deaths[identifier]
+        if deaths and (game_end is None or deaths[-1].datetime < game_end):
+            return dt_to_timestamp(deaths[-1].datetime)
+        else:
+            return dt_to_timestamp(game_end) + self._score(identifier)
+
     def get_score(self, a: Assassin) -> float:
-        return self._score(a.identifier)
\ No newline at end of file
+        return self._score(a.identifier)
+
+    # used for stats page
+    def get_conkers(self, a: Assassin) -> int:
+        return self._conkers(a.identifier)
+
+    def get_kills(self, a: Assassin) -> int:
+        return self._kills(a.identifier)
+
+    def get_attempts(self, a: Assassin) -> int:
+        return self._attempts(a.identifier)
+
+    def get_death_events(self, a: Assassin) -> List:
+        return self.deaths[a.identifier]
+
+    def get_rating(self, a: Assassin, game_end: Optional[datetime.datetime] = None) -> float:

```

Comment:
also, why do we have this indirection? why not just have `_rating` take in the assassin directly and use `a.identifier` instead of `identifier`? 

```suggestion
    def get_rating(self, a: Assassin) -> float:

```

Reply:
The purpose of the indirection is to be able to cache the value of `_rating`, which seemed reasonable at first because it was calling `_conkers` (which isn't cached) rather than `_score`
but yes the indirection is a bit silly as it is now.
I'll remove `_rating` and just put the code into `get_rating`. I would get rid of `_kills`, `_attempts` and `_bonus` but that's complicated by the fact they're used in tests.